README.md

Ponder
------

Linux & OSX: [![Build Status](https://travis-ci.org/billyquith/ponder.svg?branch=master)](https://travis-ci.org/billyquith/ponder) -
Windows: [![Build status](https://ci.appveyor.com/api/projects/status/spskn9y93e8osve2/branch/master?svg=true)](https://ci.appveyor.com/project/billyquith/ponder/branch/master)

Currents status: 3.2-alpha. API is unstable as features added/changed.

### New: Version 3

V1 replaced Boost with C++11. V2 added Lua bindings. V3 refactored to remove some warts,
ease future development, and re-add serialisation. API is evolving.

### About

Ponder is a C++ multi-purpose reflection library. It provides an abstraction for most
of the high-level concepts of C++: classes, enumerations, functions, properties.

C++1x supplies better support for compile-time reflection. Runtime reflection involves
baking compile-time reflection into the executable so this can be used at runtime. Ponder
presents a simple API for baking your objects so that they can be used at runtime. The baking
can also, optionally, userdata and support for Lua scripting.

Features:

- API to expose C++ objects.
- Runtime API to create objects, call functions, and read and modify properties.
- Automatic Lua binding generation.

### Links

 - [Homepage & documentation](http://billyquith.github.io/ponder/).
 - [GitHub project](https://github.com/billyquith/ponder).
 - [Release notes][rel].
 - [Maintainer blog](https://chinbilly.blogspot.com/search/label/ponder)

### Build

Compilers supported. [Requires C++17][compsupp].

- MSVC 2017+
- Xcode 8+
- GCC 7+
- Clang 3.9+

Ponder uses CMake to generate project files which are then used to build the project ([Build
docs][build]). E.g. Use `make` to build Ponder (but you can use any build system CMake supports):

```bash
git clone https://github.com/billyquith/ponder.git
cd ponder
mkdir build && cd build
cmake -G "Unix Makefiles" ..
make
```

### Package Managers

Windows [vcpkg][vcpkg]:

    vcpkg install ponder

### History

Ponder is a fork of [CAMP][camp], which has been retired by the original authors. CAMP relies
on Boost, and is pre-C++11. In Ponder, the Boost dependency has been removed, and instead,
C++11 features are used. CAMP was developed by Technogerma Systems France and then by [Tegesoft][tege].

See [CHANGELOG.md](CHANGELOG.md) and [release notes][rel] for more details on Ponder changes.

[rel]: https://billyquith.github.io/ponder/blog_ponder_3.html
[cl]: https://github.com/billyquith/ponder/CHANGELOG.md
[build]: https://billyquith.github.io/ponder/building_ponder.html
[tege]: http://www.tegesoft.com
[camp]: https://github.com/tegesoft/camp
[vcpkg]: https://docs.microsoft.com/en-us/cpp/build/vcpkg?view=msvc-160
[compsupp]: https://en.cppreference.com/w/cpp/compiler_support


CHANGELOG.md

Ponder Changelog
----------------

- See [release notes](https://billyquith.github.io/ponder/blog_ponder_3.html) in documentation for more details.
- See Github project issues (#nnn) for discussion on an issue.

### 3.2

- Ponder requires C++17.
- MSVC: Visual Studio 2019 support.
- Add pass-by-reference parameters (#109).
- Serialisation: JSON support added (#43).
- UserObject automatically destructs (#50).
- Big refactor to simplify property bindings & add return internal refs.
- API improvements (#93).
- Tests: Catch2 updated to 2.13.1

## 3.1

- Ponder now requires full C++14 support.
- MSVC 2017 required. MSVC 2015 does not fully support C++14 (#113).
- Fix object create with renamed classes (#108).
- Change internal class registration to by type instead of by name.
- Make detail::string_view more C++14.
- Tidy unit tests.

### 3.0.2

- Fix UserProperty creation (#110).
- Tidy UserProperty and tests.

### 3.0.1

- Fix Lua binding (#104).
- Fix `string_view` no return value (#106).
- Added docs on building Ponder.

## 3.0

- Extensive refactoring of the API.
	- Changed UserObject reference semantics. Pointer makes a reference.
- All types have common ancestor: Type.
- Added UserData per Type. (#98)
- Refactored serialisation. Removed ponder-xml. (#43).
- Added RapidJSON for use by JSON serialiser (#43).
- Added RapidXML for use by XML serialiser (#43).
- Remove dynamic read/write attribute from properties. (#91)
- Remove TagHolder. (#92)
- Removed `ponder::classByIndex()`. Use `ponder::classIterator()` instead.
- Cmake project name changed to "ponder" to match namespace (#95).
- Removed composed property getter. Use lambdas instead. See property test p14-p17.

### 2.2

- Removed dependency on cppformat (#59)

#### 2.1.1

- Identifiers/names now return IdReturn (`const std::string&`) to improve usability.

### 2.1

- Bind Lua argument convertion directly to function. Saves Value conversion.
- Lua custom type conversion with LuaValueRead and LuaValueWrite.
- Lua supports enums.
- Lua can return multiple values via `std::tuple<...>`.
- Lua functions can parse table arguments.

## 2.0

- Version 1.4 was unreleased and morphed into 2.0 because of extensive design changes.
- Runtime functionality (object factory and function calling) decoupled (#49).
- Many things renamed in the hope that this will clarify things.
  - FunctionType now FunctionKind. Avoid clash with "type".
  - ValueType now ValueKind. Avoid clash with "type".
  - Arguments now parameters. Functions have parameters and are called with arguments.
- Added support for binding to static functions (#46).
- Lua binding generator. Functional but sparse functionality. (#40).
- `PONDER_RTTI` renamed to `PONDER_POLYMORPHIC` (#36).
- Added function call return policies (#51).
- Removed UserObject parent-child. Simplification, unused.
- Removed Function callable behaviour. Unused complication.
- Class::create() allows object creation without an Args list.
- Class::tryFunction() and Class::tryProperty() added.
- Remove properties with 3 accessors. Simplification, use lambdas.


### 1.3

- Ids/Names now configurable using `IdTrait`s.
- Default `IdTrait` uses `string_view` so is more efficient (#11).
- Lambda functions recognised as class functions.
- `std::auto_ptr` removed as deprecated in C++11 and causes warning spam in GCC.
- Internal string_view class added as std one still experimental.
- Function and object classifiers.
- Bug fixes.

### 1.2

- Properties now support method chaining or named parameter idiom (#28).
- UserObjects can be created from user pointers (#30).
- Change `ponder::Type` to enum class `ValueType` (#26).
- Remove composed function bindings. Use lambdas instead.
- Re-add function unit tests.

### 1.1

- Placement new creation of objects added.
- Class function and property iterators added.
- Class function and property access by index added.
- Ability to undeclare Classes and Enums.
- Dictionaries ordered internally. Allows for fast binary searching of keys.
- Documentation fixes.

#### 1.0.2

- Moved to Catch for unit testing.

## 1.0

- Boost dependency removed. C++11 features used instead.


index.js
var index =
[
    [ "Overview", "index.html#overview", null ],
    [ "Quick example", "index.html#intro_example", null ],
    [ "Examples", "index.html#examples", null ],
    [ "Useful links", "index.html#links", null ],
    [ "Articles", "index.html#blog", null ],
    [ "License", "index.html#license", null ],
    [ "Shapes Example", "eg_page_shapes.html", null ],
    [ "Declaration Example", "eg_page_declare.html", null ],
    [ "Ponder 1.x Release Notes", "blog_ponder_1.html", [
      [ "1.0 Release Notes", "blog_ponder_1.html#ponder_1_0", null ],
      [ "1.1 Release Notes", "blog_ponder_1.html#ponder_1_1", [
        [ "Placement new creation of objects added", "blog_ponder_1.html#placement_new", null ],
        [ "Ability to undeclare Classes and Enums", "blog_ponder_1.html#undeclare", null ],
        [ "Class function and property iterators added", "blog_ponder_1.html#member_iterators", null ],
        [ "UserObject function and property access by index added", "blog_ponder_1.html#access_by_index", null ],
        [ "Dictionaries", "blog_ponder_1.html#internal_dict", null ]
      ] ],
      [ "1.2 Release Notes", "blog_ponder_1.html#ponder_1_2", [
        [ "Properties support method chaining", "blog_ponder_1.html#prop_chaining", null ],
        [ "UserObjects can be created from user pointers", "blog_ponder_1.html#userobj_pointers", null ],
        [ "Remove composed function bindings", "blog_ponder_1.html#remove_composed", null ],
        [ "ponder::Type now an enum class", "blog_ponder_1.html#enum_type", null ],
        [ "Unit tests", "blog_ponder_1.html#unittests_readded", null ]
      ] ],
      [ "1.3 Release Notes", "blog_ponder_1.html#ponder_1_3", [
        [ "Ids/Names configurable using traits", "blog_ponder_1.html#id_traits", null ],
        [ "string_view class added", "blog_ponder_1.html#string_view_added", null ],
        [ "Lambda functions recognised", "blog_ponder_1.html#lambda_added", null ],
        [ "std::auto_ptr removed", "blog_ponder_1.html#auto_ptr_removed", null ]
      ] ]
    ] ],
    [ "Ponder 2.x Release Notes", "blog_ponder_2.html", [
      [ "2.0 Release Notes", "blog_ponder_2.html#ponder_2_0", [
        [ "Decoupling", "blog_ponder_2.html#decoupling", null ],
        [ "Renaming", "blog_ponder_2.html#renaming_2", null ],
        [ "Lua binding generator", "blog_ponder_2.html#lua_support", null ],
        [ "Added support for binding to static functions (issue #46)", "blog_ponder_2.html#static_funcs", null ],
        [ "Added function call return policies (issue #51)", "blog_ponder_2.html#call_return_policy", null ],
        [ "Removed UserObject parent-child. Simplification, unused.", "blog_ponder_2.html#parent_child_removed", null ]
      ] ],
      [ "2.1 Release Notes", "blog_ponder_2.html#ponder_2_1", [
        [ "Lua: Improved function bindings", "blog_ponder_2.html#lua_binding_improved", null ],
        [ "Lua: Custom type conversion", "blog_ponder_2.html#lua_custom_type_conv", null ],
        [ "Lua: Table parameters", "blog_ponder_2.html#lua_table_params", null ],
        [ "Lua: Multiple return values", "blog_ponder_2.html#lua_multiple_return", null ],
        [ "Lua: Enum support", "blog_ponder_2.html#lua_enums", null ]
      ] ],
      [ "2.2 Release Notes", "blog_ponder_2.html#ponder_2_2", [
        [ "fmt library removed", "blog_ponder_2.html#fmt_removed", null ]
      ] ]
    ] ],
    [ "Ponder 3.x Release Notes", "blog_ponder_3.html", [
      [ "3.0 Release Notes", "blog_ponder_3.html#ponder_3_0", [
        [ "Extensive refactoring of the API.", "blog_ponder_3.html#refactor3", null ],
        [ "Changed UserObject reference semantics.", "blog_ponder_3.html#userobj_ref_semantics", null ],
        [ "All types have common ancestor: Type.", "blog_ponder_3.html#common_ancestor_type", null ],
        [ "Added UserData per Type. (#98)", "blog_ponder_3.html#userdata_added", null ],
        [ "Refactored serialisation. Removed ponder-xml. (#43).", "blog_ponder_3.html#serialisation_refactor", null ],
        [ "dynamic read/write attribute from properties. (#91)", "blog_ponder_3.html#Remove", null ],
        [ "Remove TagHolder. (#92)", "blog_ponder_3.html#tagholder_removed", null ],
        [ "Removed ponder::classByIndex().", "blog_ponder_3.html#classbyindex_removed", null ],
        [ "Cmake project name changed to \"ponder\" to match namespace (#95).", "blog_ponder_3.html#cmake_project_rename", null ],
        [ "Removed composed property getter", "blog_ponder_3.html#composed_prop_getter_removed", null ]
      ] ],
      [ "3.1 Release Notes", "blog_ponder_3.html#ponder_3_1", [
        [ "Change internal class registration to by type instead of by name", "blog_ponder_3.html#obj_reg_by_typeid", null ],
        [ "Drop support for MSVC 2015.", "blog_ponder_3.html#drop_msvc2015", null ]
      ] ],
      [ "3.2 Release Notes", "blog_ponder_3.html#ponder_3_2", null ]
    ] ]
];